File: python/horus/utils/spinner.py
import sys
import time
import threading

class Spinner:
    def __init__(self, message="Loading", delay=0.1, style="default"):
        self.message = message
        self.delay = delay
        self.running = False
        self.thread = None
        self.style = style
        
        # Different spinner styles
        self.spinner_styles = {
            'default': ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'],
            'dots': ['⠈', '⠐', '⠠', '⢀', '⡀', '⠄', '⠂', '⠁'],
            'circle': ['◐', '◓', '◑', '◒'],
            'arrows': ['←', '↖', '↑', '↗', '→', '↘', '↓', '↙'],
            'pulse': ['●', '●', '●', '○', '○', '○']
        }
        
        self.spinner_chars = self.spinner_styles.get(style, self.spinner_styles['default'])
        self.current_char = 0
    
    def _spin(self):
        """Internal spinning animation"""
        while self.running:
            char = self.spinner_chars[self.current_char]
            sys.stdout.write(f"\r  \033[96m{char}\033[0m {self.message}...")
            sys.stdout.flush()
            self.current_char = (self.current_char + 1) % len(self.spinner_chars)
            time.sleep(self.delay)
    
    def start(self):
        """Start the spinner"""
        if not self.running:
            self.running = True
            self.thread = threading.Thread(target=self._spin)
            self.thread.daemon = True
            self.thread.start()
    
    def stop(self):
        """Stop the spinner and clear the line"""
        if self.running:
            self.running = False
            if self.thread:
                self.thread.join()
            # Clear the line
            sys.stdout.write(f"\r{' ' * (len(self.message) + 10)}\r")
            sys.stdout.flush()

File: rust/src/utils/spinner.rs
use std::io::Write;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread::{self, JoinHandle};
use std::time::Duration;

pub struct Spinner {
    message: String,
    delay: Duration,
    style: Vec<&'static str>,
    running: Arc<AtomicBool>,
    handle: Option<JoinHandle<()>>,
}

impl Spinner {
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
            delay: Duration::from_millis(100),
            style: vec!["-", "\\", "|", "/"],
            running: Arc::new(AtomicBool::new(false)),
            handle: None,
        }
    }

    pub fn start(&mut self) {
        if self.running.swap(true, Ordering::Relaxed) {
            return;
        }
        let running = Arc::clone(&self.running);
        let message = self.message.clone();
        let style = self.style.clone();
        let delay = self.delay;
        self.handle = Some(thread::spawn(move || {
            let mut idx = 0usize;
            while running.load(Ordering::Relaxed) {
                print!("\r  \x1b[96m{}\x1b[0m {}...", style[idx % style.len()], message);
                let _ = std::io::stdout().flush();
                idx = idx.wrapping_add(1);
                thread::sleep(delay);
            }
        }));
    }

    pub fn stop(&mut self) {
        self.running.store(false, Ordering::Relaxed);
        if let Some(handle) = self.handle.take() {
            let _ = handle.join();
        }
        print!("\r{}\r", " ".repeat(self.message.len() + 16));
        let _ = std::io::stdout().flush();
    }
}
